### ISSUE TITLE
Panic with Repository::rev_parse_single("^")

### ISSUE THREAD

**Issue by jpgrayson** [Byron/gitoxide#706]
### Duplicates

- [X] I have searched the existing issues

### Current behavior 😯

Calling `Repository::rev_parse_single("^") causes a panic.

```
thread 'main' panicked at 'set by parser', /home/jpgrayson/code/gitoxide/git-repository/src/revision/spec/parse/delegate/mod.rs:90:71
stack backtrace:
   0: rust_begin_unwind
             at /rustc/fc594f15669680fa70d255faec3ca3fb507c3405/library/std/src/panicking.rs:575:5
   1: core::panicking::panic_fmt
             at /rustc/fc594f15669680fa70d255faec3ca3fb507c3405/library/core/src/panicking.rs:64:14
   2: core::panicking::panic_display
             at /rustc/fc594f15669680fa70d255faec3ca3fb507c3405/library/core/src/panicking.rs:135:5
   3: core::panicking::panic_str
             at /rustc/fc594f15669680fa70d255faec3ca3fb507c3405/library/core/src/panicking.rs:119:5
   4: core::option::expect_failed
             at /rustc/fc594f15669680fa70d255faec3ca3fb507c3405/library/core/src/option.rs:1879:5
   5: core::option::Option<T>::expect
             at /rustc/fc594f15669680fa70d255faec3ca3fb507c3405/library/core/src/option.rs:741:21
   6: git_repository::revision::spec::parse::delegate::<impl git_repository::revision::spec::parse::Delegate>::into_rev_spec::kind_to_spec
             at /home/jpgrayson/code/gitoxide/git-repository/src/revision/spec/parse/delegate/mod.rs:90:65
   7: git_repository::revision::spec::parse::delegate::<impl git_repository::revision::spec::parse::Delegate>::into_rev_spec
             at /home/jpgrayson/code/gitoxide/git-repository/src/revision/spec/parse/delegate/mod.rs:108:20
   8: git_repository::revision::spec::parse::<impl git_repository::revision::Spec>::from_bstr
             at /home/jpgrayson/code/gitoxide/git-repository/src/revision/spec/parse/mod.rs:38:23
   9: git_repository::repository::revision::<impl git_repository::types::Repository>::rev_parse
             at /home/jpgrayson/code/gitoxide/git-repository/src/repository/revision.rs:12:9
  10: git_repository::repository::revision::<impl git_repository::types::Repository>::rev_parse_single
             at /home/jpgrayson/code/gitoxide/git-repository/src/repository/revision.rs:28:9
```

There is an assumption in `kind_to_spec()` that a valid object preceded the `^`, but this is clearly an invalid assumption.

### Expected behavior 🤔

A plain `^` should cause an error, not a panic.

For reference, `git show ^` yields the error `fatal: bad revision '^'`.

I haven't invested enough time in the revision parsing code to suggest an approach to repairing this issue. One immediate question, however, is why revision parsing isn't using a `nom`-based approach as is used for other parsing needs in gitoxide?

### Steps to reproduce 🕹

Here is a test case that triggers the issue:

```diff
diff --git a/git-repository/tests/revision/spec/from_bytes/traverse.rs b/git-repository/tests/revision/spec/from_bytes/traverse.rs
index 6ae8b295c..a008fd18d 100644
--- a/git-repository/tests/revision/spec/from_bytes/traverse.rs
+++ b/git-repository/tests/revision/spec/from_bytes/traverse.rs
@@ -50,3 +50,9 @@ fn ancestors() {
         "Commit 55e825e has 3 ancestors along the first parent and ancestor number 42 is out of range"
     );
 }
+
+#[test]
+fn just_a_caret() {
+    let repo = repo("complex_graph").unwrap();
+    assert!(parse_spec("^", &repo).is_err());
+}
```

**Comment by Byron** [comment]
A fix will be happening very soon :D!

Can you believe that the rev-spec parser has been fuzzed? And that the fuzzer didn't find this one? And that I was convinced that it thus must be pretty perfect (after fixing the things it did find) ;)? Amazing find for sure!

**Comment by Byron** [comment]
This has been fixed via #707 and the issue was ultimately caused by an invalid assumption just like you stated.

> One immediate question, however, is why revision parsing isn't using a nom-based approach as is used for other parsing needs in gitoxide?

With `nom` my brain convulses and I would have trouble doing anything that is a little more flexible and complicated, like rev-spec parsing ought to be. `nom` is nice to encode 'patterns' which usually yields quite readable code while performing well enough. Unfortunately I never really warmed up to it and feel I have to re-learn it next time I use it. It's a sad affair of the "it's me, not you" kind I think.

**Comment by jpgrayson** [comment]
Thanks for the quick action on this, @Byron. Your repair seems much more comprehensive than what I was looking at.
